decision_service/dsclientlib/logger/ds_async_batcher.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "ds_concurrent_queue.h"
+#include "ds_api_status.h"
+
+#include <memory>
+#include <string>
+#include "../error_callback_fn.h"
+#include <thread>
+
+namespace decision_service {
+  class error_callback_fn;
+
+	// This class takes uses a queue and a background thread to accumulate events, and send them by batch asynchronously.
+	// A batch is shipped with TSender::send(data)
+	template <typename TSender>
+	class async_batcher {
+
+	public:
+
+		int append(const std::string& evt, api_status* status = nullptr)
+		{
+			if (_queue.size() < _queue_max_size) {
+				_queue.push(evt);
+				return error_code::success;
+			}
+
+			//report errors
+			api_status::try_update(status, error_code::background_queue_overflow, "dropped event: " + evt);
```

Comment:
>api_status::try_update(status, error_code::background_queue_overflow, "dropped event: " + evt); [](start = 3, length = 95)

at least in the past we adjusted the probabilities if we're dropping events.

Reply:
add work item and schedule

---
In reply to: [186742634](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186742634) [](ancestors = 186742634)

Reply:
Added work item under logging.

---
In reply to: [187653180](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r187653180) [](ancestors = 187653180,186742634)

Reply:
Done.

---
In reply to: [187774785](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r187774785) [](ancestors = 187774785,187653180,186742634)